comictaggerlib/settingswindow.py

```diff
@@ -183,14 +185,78 @@ def __init__(self, parent: QtWidgets.QWidget, settings: ComicTaggerSettings) ->
+            # Use a dict to make a var name from var
+            temp = {}
+            tab = source["talker"].source_details["id"]
+            temp[tab] = QtWidgets.QWidget()
+
+            # Add new sub tab to Comic Source tab
+            self.tComicSourcesOptions.addTab(temp[tab], source["talker"].source_details["name"])
+            layout = QtWidgets.QVBoxLayout(self)
+            layout_grid = QtWidgets.QGridLayout(self)
+
+            row = 0
+            for option in source["talker"].source_details["settings_options"]:
+                option_name = source["talker"].source_details["id"] + "_" + option["name"]
+
+                if option["type"] == "bool":
+                    # bool equals a checkbox (QCheckBox)
+                    setattr(self, option_name, QtWidgets.QCheckBox(option["text"]))
+                    # Set widget status
+                    # This works because when a talker class is initialised it loads its settings from disk
+                    if source["talker"].options[option["name"]]:
+                        getattr(self, option_name).setChecked(source["talker"].options[option["name"]])
+                    ## Add widget to VBoxlayout
+                    layout.addWidget(getattr(self, option_name))
+                if option["type"] == "str":
+                    # str equals a text field (QLineEdit)
+                    lbl_name = "le" + option_name
+                    # Create a label
+                    setattr(self, lbl_name, QtWidgets.QLabel(option["text"]))
+                    layout_grid.addWidget(getattr(self, lbl_name), row, 0)
+                    setattr(self, option_name, QtWidgets.QLineEdit())
+                    # Set widget status
+                    if source["talker"].options[option["name"]]:
+                        getattr(self, option_name).setText(
+                            source["talker"].options[option["name"]])
+                    layout_grid.addWidget(getattr(self, option_name), row, 1)
+                    # Special case for api_key, make a test button
+                    if option["name"] == "api_key":
+                        tbtn_name = "btn" + option_name
+                        setattr(self, tbtn_name, QtWidgets.QPushButton("Test Key"))
+                        layout_grid.addWidget(getattr(self, tbtn_name), row, 2)
+                        getattr(self, tbtn_name).clicked.connect(
+                            lambda: self.test_api_key(source["talker"].source_details["id"])
+                        )
+
+                    row += 1
+
+                # Add tooltip text
+                getattr(self, option_name).setToolTip(option["help_text"])
+
+            # Add vertical spacer
+            vspacer = QtWidgets.QSpacerItem(20, 40, QtWidgets.QSizePolicy.Minimum, QtWidgets.QSizePolicy.Expanding)
+            layout_grid.addItem(vspacer, row, 0)
+            layout.addLayout(layout_grid)
+            # Display the new widgets
+            temp[tab].setLayout(layout)
+
```

Comment:
```suggestion
        self.sources_api = SourcesAPI()
        self.generate_source_option_tabs()

    def generate_source_option_tabs(self) -> None:
        # Add source sub tabs to Comic Sources tab
        for source in self.sources_api.sources:
            # TODO Add source to general tab dropdown list. Option does nothing currently.
            self.cobxInfoSource.addItem(source["talker"].source_details["name"], source["talker"].source_details["id"])

            # Use a dict to make a var name from var
            source_info = {}
            tab_name = source["talker"].source_details["id"]
            source_info[tab_name] = {"tab": QtWidgets.QWidget(), "widgets": {}}

            layout_grid = QtWidgets.QGridLayout()

            row = 0
            for option in source["talker"].source_details["settings_options"].values():
                option_name = source["talker"].source_details["id"] + "_" + option["name"]
                current_widget = None

                if option["type"] is bool:
                    # bool equals a checkbox (QCheckBox)
                    current_widget = QtWidgets.QCheckBox(option["text"])

                    # Set widget status
                    # This works because when a talker class is initialised it loads its settings from disk
                    if option["value"]:
                        current_widget.setChecked(option["value"])
                    # Add widget to VBoxlayout
                    layout_grid.addWidget(current_widget, row, 0)
                if option["type"] is str:
                    # str equals a text field (QLineEdit)
                    lbl = QtWidgets.QLabel(option["text"])
                    # Create a label
                    # setattr(self, "le" + option_name, lbl)
                    layout_grid.addWidget(lbl, row, 0)

                    current_widget = QtWidgets.QLineEdit()

                    # Set widget status
                    if option["value"]:
                        current_widget.setText(option["value"])
                    layout_grid.addWidget(current_widget, row, 1)

                    # Special case for api_key, make a test button
                    if option["name"] == "api_key":
                        btn = QtWidgets.QPushButton("Test Key")

                        layout_grid.addWidget(btn, row, 2)
                        btn.clicked.connect(lambda: self.test_api_key(source["talker"].source_details["id"]))
                        # setattr(self, "btn" + option_name, btn)

                row += 1
                # Add tooltip text
                if current_widget:
                    current_widget.setToolTip(option["help_text"])

                source_info[tab_name]["widgets"][option["name"]] = current_widget

            # Add vertical spacer
            vspacer = QtWidgets.QSpacerItem(20, 40, QtWidgets.QSizePolicy.Minimum, QtWidgets.QSizePolicy.Expanding)
            layout_grid.addItem(vspacer, row, 0)
            # layout.addLayout(layout_grid)
            # Display the new widgets
            source_info[tab_name]["tab"].setLayout(layout_grid)

            # Add new sub tab to Comic Source tab
            self.tComicSourcesOptions.addTab(source_info[tab_name]["tab"], source["talker"].source_details["name"])

            self.sources = source_info

```

Reply:
The reason for the two layouts was so that check boxes would be grouped together independent of where they came in the list but it seems reasonable to expect the settings_options list to be in the required order?

Reply:
Letting the source determine the order also allows someone to have an enable check box right before a value